platform/linux-generic/test/validation/api/ml/ml_linux.c

```diff
@@ -0,0 +1,1259 @@
+ * Copyright (c) 2023 Nokia
+ */
+
+#ifndef _GNU_SOURCE
+#define _GNU_SOURCE
+#endif
+
+#include <unistd.h>
+#include <string.h>
+#include <libgen.h>
+#include <odp_api.h>
+#include <odp/helper/odph_api.h>
+#include "odp_cunit_common.h"
+
+#define TIMEOUT		5
+#define MODEL_NAME	"Test"
+#define NUM_INPUTS	1
+#define NUM_OUTPUTS	1
+#define RUN_NUM		2
+#define BUF_LEN		256
+#define MAX_COMPL_ID	20
+#define CONFIG_MAX_MODEL_SIZE	500
+
+#define COMPL_POOL_NAME "ML compl pool"
+#define NUM_COMPL	10
+
+/**
+ * About model simple_linear.onnx being tested in this suite
+ *
+ * Model info:
+ *	Version: 1
+ *	Inputs: name: x, type: int32, shape: [1]
+ *	Outputs: name: y, type: int32, shape: [1]
+ *
+ * The model is of form y = 3 * x + 4
+ * Thus when x = 5, the output y should be 19.
+ */
+typedef struct global_t {
+	int disabled;
+	odp_ml_capability_t ml_capa;
+	odp_ml_config_t ml_config;
+	odp_ml_model_param_t model_param;
+	odp_ml_model_t ml_model;
+	odp_pool_t compl_pool;
+	odp_queue_t queue;
+	odp_ml_data_t data;
+	odp_ml_data_seg_t input_seg;
+	odp_ml_data_seg_t output_seg;
+	odp_ml_run_param_t run_param;
+	uint64_t wait_ns;
+	int32_t x;
+	int32_t y;
+	int32_t y_expected;
+
+} global_t;
+
+static global_t global;
+
+static int fill_model_param(const char *model_name, odp_ml_model_param_t *model_param)
+{
+	size_t size;
+	char *pos;
+	char *exe_dir;
+	size_t exe_dir_len;
+	FILE *model_file;
+	char exe_path[BUF_LEN];
+	ssize_t exe_path_len;
+	char model_path[BUF_LEN];
+
+	/* Model file is placed in the same directory as the executable ml_linux */
+	exe_path_len = readlink("/proc/self/exe", exe_path, BUF_LEN - 1);
+	if (exe_path_len != -1) {
+		exe_path[exe_path_len] = '\0';
+
+		pos = strstr(exe_path, ".libs");
+		if (pos)
+			*(pos + 5) = '\0';
+
+		exe_dir = dirname(exe_path);
+		exe_dir_len = strlen(exe_dir);
+
+		memcpy(model_path, exe_dir, exe_dir_len);
+		model_path[exe_dir_len] = '/';
+		model_path[exe_dir_len + 1] = '\0';
+
+		strncat(model_path, model_name, BUF_LEN - strlen(model_path) - 1);
+		ODPH_DBG("model_path: %s\n", model_path);
+		model_file = fopen(model_path, "rb");
+	} else { /* Can't get executable path, try to find model file at current dir*/
+		model_file = fopen(model_name, "rb");
+	}
+
+	if (model_file == NULL) {
+		perror("Failed to open model file");
+		return -1;
+	}
+
+	/* Get the model file size in bytes */
+	fseek(model_file, 0, SEEK_END);
+	model_param->size = ftell(model_file);
+	rewind(model_file);
+
+	model_param->model = malloc(model_param->size);
+	if (!model_param->model) {
+		ODPH_ERR("\n\nMemory allocation failed\n");
+		fclose(model_file);
+		return -1;
+	}
+	size = fread(model_param->model, model_param->size, 1, model_file);
+
+	fclose(model_file);
+	if (size != 1) {
+		ODPH_ERR("\n\nRead model file failed\n");
+		return -1;
+	}
+
+	if ((global.ml_config.load_mode_mask & ODP_ML_COMPL_MODE_POLL) ||
+	    (global.ml_config.run_mode_mask & ODP_ML_COMPL_MODE_POLL))
+		model_param->max_compl_id = MAX_COMPL_ID;
```

Comment:
It seems that max compl ID capa is not checked (max is 20, value of 4 is used in test). If only one ID is needed, value of 0 could be used without capa check.

Reply:
Using compl id 0 in v3.